airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java

```diff
@@ -52,28 +55,33 @@
   private final Source<AirbyteMessage> source;
   private final Mapper<AirbyteMessage> mapper;
   private final Destination<AirbyteMessage> destination;
-  private final MessageTracker<AirbyteMessage> messageTracker;
+  private final MessageTracker<AirbyteMessage> sourceMessageTracker;
+  private final MessageTracker<AirbyteMessage> destinationMessageTracker;
 
   private final AtomicBoolean cancelled;
+  private final AtomicBoolean hasFailed;
 
   public DefaultReplicationWorker(final String jobId,
                                   final int attempt,
                                   final Source<AirbyteMessage> source,
                                   final Mapper<AirbyteMessage> mapper,
                                   final Destination<AirbyteMessage> destination,
-                                  final MessageTracker<AirbyteMessage> messageTracker) {
+                                  final MessageTracker<AirbyteMessage> sourceMessageTracker,
+                                  final MessageTracker<AirbyteMessage> destinationMessageTracker) {
     this.jobId = jobId;
     this.attempt = attempt;
     this.source = source;
     this.mapper = mapper;
     this.destination = destination;
-    this.messageTracker = messageTracker;
+    this.sourceMessageTracker = sourceMessageTracker;
+    this.destinationMessageTracker = destinationMessageTracker;
 
     this.cancelled = new AtomicBoolean(false);
+    this.hasFailed = new AtomicBoolean(false);
   }
 
   @Override
-  public StandardSyncOutput run(StandardSyncInput syncInput, Path jobRoot) throws WorkerException {
+  public ReplicationOutput run(StandardSyncInput syncInput, Path jobRoot) throws WorkerException {
```

Comment:
nit: I think it would be helpful to summarise `run` behaviour about state emission and saving state. i.e. State remains emitted by the source and is only saved in the Destination. Otherwise I can see someone getting confused about where state originates from.